File: crates/kuben-api/src/routes/environments.rs
//! Environments of a project, on the SQL model (ADR-032). SQL holds them
//! and their placement; the materializer writes the `Environment` resource
//! right away (its controller provisions the namespace) and removes it on
//! deletion. The projection adds live status.

use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use kuben_core::{Error, perm::Perm};
use kuben_crd::Quota;
use kuben_platform::{controller::resources::namespace_name, projection::EnvironmentView};
use kuben_store::repo::{ENVIRONMENT_APPLY, ENVIRONMENT_DELETE, EnvironmentKind, EnvironmentRecord, Subject};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::{request, scope, validate};
use crate::{authz::Authz, error::ApiResult, state::ApiState};

/// Grace period before a deleted production environment is purged.
pub const PRODUCTION_DELETION_GRACE: &str = "168h";
/// The cluster every placement is on until multi-cluster (M1.9 and later).
const PRIMARY_CLUSTER: &str = "primary";

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "lowercase")]
pub enum EnvType {
    #[default]
    Standard,
    Production,
    Preview,
}

impl From<EnvType> for EnvironmentKind {
    fn from(t: EnvType) -> Self {
        match t {
            EnvType::Standard => Self::Standard,
            EnvType::Production => Self::Production,
            EnvType::Preview => Self::Preview,
        }
    }
}

#[derive(Debug, Serialize, ToSchema)]
pub struct EnvironmentDto {
    /// Short name used in URLs, e.g. `prod`.
    pub name: String,
    /// Kubernetes object name, e.g. `shop-prod`.
    pub resource_name: String,
    pub project: String,
    /// `standard`, `production` or `preview`.
    pub env_type: String,
    pub namespace: String,
    /// `Pending`, `Ready`, `Terminating` or `Degraded`.
    pub phase: Option<String>,
    pub ready: bool,
    pub message: Option<String>,
    pub deleting: bool,
    /// When a soft-deleted environment will be purged.
    pub deletion_scheduled_at: Option<String>,
    pub created_at: Option<String>,
}

impl EnvironmentDto {
    #[must_use]
    pub fn of(project: &str, e: &EnvironmentRecord, view: Option<&EnvironmentView>) -> Self {
        let resource_name = scope::environment_resource_name(project, &e.slug);
        let phase = view
            .and_then(|v| v.phase.clone())
            .or_else(|| Some(if e.deleting { "Terminating" } else { "Pending" }.to_owned()));
        Self {
            name: e.slug.clone(),
            namespace: e
                .namespace
                .clone()
                .unwrap_or_else(|| namespace_name(&resource_name)),
            resource_name,
            project: project.to_owned(),
            env_type: e.env_type.clone(),
            phase,
            ready: view.is_some_and(|v| v.ready) && !e.deleting,
            message: view.and_then(|v| v.message.clone()),
            deleting: e.deleting,
            deletion_scheduled_at: view.and_then(|v| v.deletion_scheduled_at.clone()),
            created_at: Some(request::timestamp(e.created_at)),
        }
    }
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct QuotaInput {
    #[schema(example = "4")]
    pub cpu: Option<String>,
    #[schema(example = "8Gi")]
    pub memory: Option<String>,
    #[schema(example = 50)]
    pub pods: Option<u32>,
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct CreateEnvironment {
    /// Short name, e.g. `prod` (the object is named `<project>-<name>`).
    #[schema(example = "prod")]
    pub name: String,
    #[serde(default)]
    pub env_type: EnvType,
    pub quota: Option<QuotaInput>,
}

/// List a project's environments.
#[utoipa::path(
    get,
    path = "/projects/{project}/environments", operation_id = "listEnvironments",
    tag = "environments",
    params(("project" = String, Path, description = "Project name")),
    responses(
        (status = 200, body = Vec<EnvironmentDto>),
        (status = 404, body = crate::error::Problem),
    )
)]
pub async fn list(
    State(state): State<ApiState>,
    authz: Authz,
    Path(project): Path<String>,
) -> ApiResult<Json<Vec<EnvironmentDto>>> {
    let p = scope::project(&state, &authz, &project).await?;
    let _proof = authz.require(&state, Perm::EnvRead, &p.chain())?;
    let mut tenant = state.store.tenant(p.org).await?;
    let items = tenant
        .environments(p.id())
        .await?
        .iter()
        .map(|e| {
            let view = state
                .projections
                .environment(&scope::environment_resource_name(p.slug(), &e.slug));
            EnvironmentDto::of(p.slug(), e, view.as_deref())
        })
        .collect();
    Ok(Json(items))
}

/// One environment.
#[utoipa::path(
    get,
    path = "/projects/{project}/environments/{environment}", operation_id = "getEnvironment",
    tag = "environments",
    params(
        ("project" = String, Path, description = "Project name"),
        ("environment" = String, Path, description = "Environment short name"),
    ),
    responses((status = 200, body = EnvironmentDto), (status = 404, body = crate::error::Problem))
)]
pub async fn get(
    State(state): State<ApiState>,
    authz: Authz,
    Path((project, environment)): Path<(String, String)>,
) -> ApiResult<Json<EnvironmentDto>> {
    let e = scope::environment(&state, &authz, &project, &environment).await?;
    let _proof = authz.require(&state, Perm::EnvRead, &e.chain())?;
    Ok(Json(EnvironmentDto::of(
        e.project.slug(),
        &e.env,
        e.view.as_deref(),
    )))
}

fn quota(input: Option<QuotaInput>) -> ApiResult<Option<serde_json::Value>> {
    let Some(q) = input else {
        return Ok(None);
    };
    if let Some(cpu) = &q.cpu {
        validate::quantity("quota.cpu", cpu)?;
    }
    if let Some(mem) = &q.memory {
        validate::quantity("quota.memory", mem)?;
    }
    let quota = Quota {
        cpu: q.cpu,
        memory: q.memory,
        pods: q.pods,
    };
    Ok(Some(serde_json::to_value(quota).map_err(Error::internal)?))
}

/// Create an environment: its namespace follows.
#[utoipa::path(
    post,
    path = "/projects/{project}/environments", operation_id = "createEnvironment",
    tag = "environments",
    params(("project" = String, Path, description = "Project name")),
    request_body = CreateEnvironment,
    responses(
        (status = 201, body = EnvironmentDto),
        (status = 403, body = crate::error::Problem),
        (status = 409, body = crate::error::Problem),
        (status = 422, body = crate::error::Problem),
    )
)]
pub async fn create(
    State(state): State<ApiState>,
    authz: Authz,
    Path(project): Path<String>,
    Json(body): Json<CreateEnvironment>,
) -> ApiResult<(StatusCode, Json<EnvironmentDto>)> {
    let p = scope::project(&state, &authz, &project).await?;
    let _proof = authz.require(&state, Perm::EnvWrite, &p.chain())?;
    validate::dns_label("name", &body.name, 20)?;
    let resource = scope::environment_resource_name(p.slug(), &body.name);
    let namespace = namespace_name(&resource);
    if namespace.len() > 63 {
        return Err(Error::Validation("project and environment names are too long together".into()).into());
    }
    let quota = quota(body.quota)?;
    if p.project.deleting {
        return Err(Error::Conflict(format!("project `{}` is being deleted", p.slug())).into());
    }
    // Environment resources are cluster-wide.
    let taken = || Error::Conflict(format!("environment `{}` already exists", body.name));
    if state
        .projections
        .environment(&resource)
        .is_some_and(|v| v.org.as_deref() != Some(p.org.to_string().as_str()))
    {
        return Err(taken().into());
    }
    let (_, actor) = request::actor(&authz);
    let what = format!("environment `{}`", body.name);
    let mut tenant = state.store.tenant(p.org).await?;
    let id = tenant
        .create_environment_typed(
            p.id(),
            &body.name,
            &body.name,
            body.env_type.into(),
            quota.as_ref(),
        )
        .await
        .map_err(|e| request::duplicate(e, &what))?;
    let cluster = tenant.ensure_cluster(PRIMARY_CLUSTER).await?;
    tenant
        .create_placement(p.id(), id, cluster, &namespace)
        .await
        .map_err(|e| request::duplicate(e, &what))?;
    tenant
        .request(
            ENVIRONMENT_APPLY,
            Subject::environment(p.id(), id),
            &actor,
            request::audit(&authz, ENVIRONMENT_APPLY, "environment", resource),
        )
        .await?;
    let env = tenant.environment(p.id(), &body.name).await?.ok_or_else(taken)?;
    tenant.commit().await?;
    Ok((
        StatusCode::CREATED,
        Json(EnvironmentDto::of(p.slug(), &env, None)),
    ))
}

/// Delete an environment. Production environments need the
/// `env-delete-protected` permission and are purged after a grace period.
#[utoipa::path(
    delete,
    path = "/projects/{project}/environments/{environment}", operation_id = "deleteEnvironment",
    tag = "environments",
    params(
        ("project" = String, Path, description = "Project name"),
        ("environment" = String, Path, description = "Environment short name"),
    ),
    responses(
        (status = 202, description = "Deletion accepted"),
        (status = 403, body = crate::error::Problem),
        (status = 404, body = crate::error::Problem),
        (status = 409, body = crate::error::Problem),
    )
)]
pub async fn delete(
    State(state): State<ApiState>,
    authz: Authz,
    Path((project, environment)): Path<(String, String)>,
) -> ApiResult<StatusCode> {
    let e = scope::environment(&state, &authz, &project, &environment).await?;
    let perm = if e.env.env_type == "production" {
        Perm::EnvDeleteProtected
    } else {
        Perm::EnvWrite
    };
    let _proof = authz.require(&state, perm, &e.chain())?;
    let mut tenant = state.store.tenant(e.project.org).await?;
    if !tenant.mark_environment_deleting(e.id()).await? {
        return Err(Error::Conflict(format!("environment `{environment}` is being deleted")).into());
    }
    let (_, actor) = request::actor(&authz);
    tenant
        .request(
            ENVIRONMENT_DELETE,
            Subject::environment(e.project.id(), e.id()),
            &actor,
            request::audit(&authz, ENVIRONMENT_DELETE, "environment", e.resource_name()),
        )
        .await?;
    tenant.commit().await?;
    Ok(StatusCode::ACCEPTED)
}

File: go/hub/internal/api/environments.go
package api

// Environments of a project, on the SQL model (routes/environments.rs,
// ADR-032). SQL holds them and their placement; the materializer writes the
// Environment resource right away (its controller provisions the namespace)
// and removes it on deletion. The projection adds live status.

import (
	"context"

	"github.com/Teamtem-dev/kuben/go/hub/internal/api/access"
	"github.com/Teamtem-dev/kuben/go/hub/internal/api/gen"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/capacity"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/kerr"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/perm"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/policy"
	"github.com/Teamtem-dev/kuben/go/hub/internal/platform/projection"
	"github.com/Teamtem-dev/kuben/go/hub/internal/platform/render"
	"github.com/Teamtem-dev/kuben/go/hub/internal/store"
)

// primaryCluster is the cluster every placement is on until multi-cluster
// (M1.9 and later).
const primaryCluster = "primary"

// environmentDto is EnvironmentDto::of.
func environmentDto(project string, e store.EnvironmentRecord, view opt.Val[projection.EnvironmentView]) gen.EnvironmentDto {
	resource := render.EnvironmentResourceName(project, e.Slug)
	v, seen := view.Get()
	phase := "Pending"
	if e.Deleting {
		phase = "Terminating"
	}
	if seen {
		phase = v.Phase.Or(phase)
	}
	message, scheduled := opt.None[string](), opt.None[string]()
	if seen {
		message, scheduled = v.Message, v.DeletionScheduledAt
	}
	return gen.EnvironmentDto{
		Name:                e.Slug,
		ResourceName:        resource,
		Project:             project,
		EnvType:             e.EnvType,
		Namespace:           e.Namespace.Or(render.NamespaceName(resource)),
		Phase:               gen.NewOptNilString(phase),
		Ready:               seen && v.Ready && !e.Deleting,
		Message:             optNilString(message),
		Deleting:            e.Deleting,
		DeletionScheduledAt: optNilString(scheduled),
		CreatedAt:           gen.NewOptNilString(Timestamp(e.CreatedAt)),
	}
}

// ListEnvironments lists a project's environments.
func (s *Server) ListEnvironments(ctx context.Context, params gen.ListEnvironmentsParams) (gen.ListEnvironmentsRes, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	p, err := s.findProject(ctx, a, params.Project)
	if err != nil {
		return nil, err
	}
	if _, err := a.Require(perm.EnvRead, p.chain()); err != nil {
		return nil, err //nolint:wrapcheck // a kerr already
	}
	t, err := s.deps.Store.Tenant(ctx, p.org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // read only
	environments, err := t.Environments(ctx, p.project.ID)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	items := gen.ListEnvironmentsOKApplicationJSON{}
	for _, e := range environments {
		view := s.environmentView(render.EnvironmentResourceName(p.project.Slug, e.Slug))
		items = append(items, environmentDto(p.project.Slug, e, view))
	}
	return &items, nil
}

// GetEnvironment is one environment.
func (s *Server) GetEnvironment(ctx context.Context, params gen.GetEnvironmentParams) (gen.GetEnvironmentRes, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	e, err := s.findEnvironment(ctx, a, params.Project, params.Environment)
	if err != nil {
		return nil, err
	}
	if _, err := a.Require(perm.EnvRead, e.chain()); err != nil {
		return nil, err //nolint:wrapcheck // a kerr already
	}
	dto := environmentDto(e.project.project.Slug, e.env, e.view)
	return &dto, nil
}

// environmentQuota checks a requested quota and returns it as stored:
// admission reads the same quantities (M4.5).
func environmentQuota(input gen.OptNilQuotaInput) (opt.Val[any], error) {
	q, ok := input.Get()
	if !ok {
		return opt.None[any](), nil
	}
	quota := map[string]any{}
	if cpu, ok := q.CPU.Get(); ok {
		if err := Quantity("quota.cpu", cpu); err != nil {
			return opt.None[any](), err
		}
		if _, ok := capacity.CPUMillis(cpu); !ok {
			return opt.None[any](), kerr.New(kerr.Validation, "quota.cpu `%s` is not a CPU quantity", cpu)
		}
		quota["cpu"] = cpu
	}
	if memory, ok := q.Memory.Get(); ok {
		if err := Quantity("quota.memory", memory); err != nil {
			return opt.None[any](), err
		}
		if _, ok := capacity.Bytes(memory); !ok {
			return opt.None[any](), kerr.New(kerr.Validation, "quota.memory `%s` is not a memory quantity", memory)
		}
		quota["memory"] = memory
	}
	if pods, ok := q.Pods.Get(); ok {
		if pods < 0 {
			return opt.None[any](), kerr.New(kerr.Validation, "quota.pods must not be negative")
		}
		quota["pods"] = pods
	}
	return opt.Some[any](quota), nil
}

// CreateEnvironment creates an environment; its namespace follows.
func (s *Server) CreateEnvironment(
	ctx context.Context, req *gen.CreateEnvironment, params gen.CreateEnvironmentParams,
) (gen.CreateEnvironmentRes, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	p, err := s.findProject(ctx, a, params.Project)
	if err != nil {
		return nil, err
	}
	if _, err := a.Require(perm.EnvWrite, p.chain()); err != nil {
		return nil, err //nolint:wrapcheck // a kerr already
	}
	if err := DNSLabel("name", req.Name, 20); err != nil {
		return nil, err
	}
	resource := render.EnvironmentResourceName(p.project.Slug, req.Name)
	namespace := render.NamespaceName(resource)
	if len(namespace) > 63 {
		return nil, kerr.New(kerr.Validation, "project and environment names are too long together")
	}
	quota, err := environmentQuota(req.Quota)
	if err != nil {
		return nil, err
	}
	if p.project.Deleting {
		return nil, kerr.New(kerr.Conflict, "project `%s` is being deleted", p.project.Slug)
	}
	// Environment resources are cluster-wide.
	taken := kerr.New(kerr.Conflict, "environment `%s` already exists", req.Name)
	if v, ok := s.deps.Projections.Environment(resource); ok && v.Org != opt.Some(p.org.String()) {
		return nil, taken
	}
	record, err := s.createEnvironment(ctx, a, p, req.Name, environmentKind(req.EnvType.Or(gen.EnvTypeStandard)), quota, namespace)
	if err != nil {
		return nil, err
	}
	dto := environmentDto(p.project.Slug, record, opt.None[projection.EnvironmentView]())
	return &dto, nil
}

// environmentKind is From<EnvType> for EnvironmentKind.
func environmentKind(t gen.EnvType) store.EnvironmentKind {
	switch t {
	case gen.EnvTypeProduction:
		return store.Production
	case gen.EnvTypePreview:
		return store.Preview
	case gen.EnvTypeStandard:
	}
	return store.Standard
}

// createEnvironment records the environment, its placement and first
// policy, and asks the materializer to write it, in one transaction.
func (s *Server) createEnvironment(
	ctx context.Context, a access.Access, p projectScope, name string, kind store.EnvironmentKind, quota opt.Val[any], namespace string,
) (store.EnvironmentRecord, error) {
	taken := kerr.New(kerr.Conflict, "environment `%s` already exists", name)
	what := "environment `" + name + "`"
	_, actor := a.Actor()
	t, err := s.deps.Store.Tenant(ctx, p.org)
	if err != nil {
		return store.EnvironmentRecord{}, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // a no-op after the commit
	if err := s.admitEnvironment(ctx, t); err != nil {
		return store.EnvironmentRecord{}, err
	}
	id, err := t.CreateEnvironmentTyped(ctx, p.project.ID, name, name, kind, quota)
	if err != nil {
		return store.EnvironmentRecord{}, duplicate(err, what)
	}
	cluster, err := t.EnsureCluster(ctx, primaryCluster)
	if err != nil {
		return store.EnvironmentRecord{}, err //nolint:wrapcheck // a store error, answered as internal
	}
	if _, err := t.CreatePlacement(ctx, p.project.ID, id, cluster, namespace); err != nil {
		return store.EnvironmentRecord{}, duplicate(err, what)
	}
	// Protection is recorded, never inferred later: production starts with
	// one approval by someone other than the requester (M4.1).
	initial := policy.Initial(kind == store.Production)
	if _, set, err := t.SetEnvironmentPolicy(ctx, p.project.ID, id, initial, actor); err != nil || !set {
		return store.EnvironmentRecord{}, orConflict(err, taken)
	}
	resource := render.EnvironmentResourceName(p.project.Slug, name)
	if _, err := t.Request(ctx, store.EnvironmentApply, store.EnvironmentSubject(p.project.ID, id), actor,
		requestAudit(a, store.EnvironmentApply, "environment", resource)); err != nil {
		return store.EnvironmentRecord{}, err //nolint:wrapcheck // a store error, answered as internal
	}
	record, found, err := t.Environment(ctx, p.project.ID, name)
	if err != nil || !found {
		return store.EnvironmentRecord{}, orConflict(err, taken)
	}
	if err := t.Commit(ctx); err != nil {
		return store.EnvironmentRecord{}, err //nolint:wrapcheck // a store error, answered as internal
	}
	return record, nil
}

// orConflict is err, or conflict when there is none.
func orConflict(err, conflict error) error {
	if err != nil {
		return err
	}
	return conflict
}

// admitEnvironment is admission::admit_environment: the organization's
// environment quota (M4.5).
func (s *Server) admitEnvironment(ctx context.Context, t *store.Tenant) error {
	return t.AdmitEnvironment(ctx, s.deps.Config.Quota.OrgEnvironments) //nolint:wrapcheck // a kerr conflict or a store error
}

// DeleteEnvironment deletes an environment. Production environments need
// the `env-delete-protected` permission and are purged after a grace
// period.
func (s *Server) DeleteEnvironment(ctx context.Context, params gen.DeleteEnvironmentParams) (gen.DeleteEnvironmentRes, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	e, err := s.findEnvironment(ctx, a, params.Project, params.Environment)
	if err != nil {
		return nil, err
	}
	need := perm.EnvWrite
	if e.env.EnvType == "production" {
		need = perm.EnvDeleteProtected
	}
	if _, err := a.Require(need, e.chain()); err != nil {
		return nil, err //nolint:wrapcheck // a kerr already
	}
	t, err := s.deps.Store.Tenant(ctx, e.project.org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // a no-op after the commit
	// A preview deleted by hand is closed: its pull request's next event
	// does not bring it back (M5.1).
	if _, err := t.ClosePreview(ctx, e.env.ID, store.CloseDeleted, s.deps.Clock.NowMs()); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	marked, err := t.MarkEnvironmentDeleting(ctx, e.env.ID)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	if !marked {
		return nil, kerr.New(kerr.Conflict, "environment `%s` is being deleted", params.Environment)
	}
	_, actor := a.Actor()
	if _, err := t.Request(ctx, store.EnvironmentDelete, store.EnvironmentSubject(e.project.project.ID, e.env.ID), actor,
		requestAudit(a, store.EnvironmentDelete, "environment", e.resourceName())); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	if err := t.Commit(ctx); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	return &gen.DeleteEnvironmentAccepted{}, nil
}
